src/inet/UDPEndPointImplLwIP.cpp

```diff
@@ -66,27 +63,63 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
 namespace chip {
 namespace Inet {
 
-namespace {
-/**
- * @brief
- * RAII locking for LwIP core to simplify management of
- * LOCK_TCPIP_CORE()/UNLOCK_TCPIP_CORE() calls.
- */
-class ScopedLwIPLock
+EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+
+err_t EndPointStateLwIP::RunOnTCPIPRet(std::function<err_t()> fn)
 {
-public:
-    ScopedLwIPLock() { LOCK_TCPIP_CORE(); }
-    ~ScopedLwIPLock() { UNLOCK_TCPIP_CORE(); }
-};
-} // anonymous namespace
+    assertChipStackLockedByCurrentThread();
+#if LWIP_TCPIP_CORE_LOCKING
+    err_t err;
+    LOCK_TCPIP_CORE();
+    err = fn();
+    UNLOCK_TCPIP_CORE();
+    return err;
+#else
+    // Post a message to the TCPIP task and wait for it to run.
+    static sys_sem_t sTCPIPSem;
+    static bool sTCPIPSemInited = false;
+    if (!sTCPIPSemInited)
+    {
+        err_t err = sys_sem_new(&sTCPIPSem, 0);
+        if (err != ERR_OK)
+        {
+            return err;
+        }
+        sTCPIPSemInited = true;
+    }
 
-EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+    static std::function<err_t()> sTCPIPFunction;
+    static err_t sTCPIPFunctionResult;
```

Comment:
I assume the goal here with these statics is to not have the lambda closing over stack variables to avoid heap-allocations?  If so, please document that.  If that's _not_ the goal, it's even more important to document what the goal is.

Reply:
the reason is that the lambda passed to `tcpip_callback()` is not really a lambda but a C function pointer. lambda syntax in this case is just syntactic sugar, it's not a real lambda and we can't really capture anything.
we could allocate a temporary struct to hold these and pass as context, i thought two small statics would be ok for our purpose, especially since we `assertChipStackLockedByCurrentThread` already and don't expect this to be invoked concurrently. i added a comment.